jython-3-roadmap.md

```diff
@@ -0,0 +1,293 @@
+
+1. AST classes generated from Python ASDL.
+Generated classes are Python objects in an `ast` module.
+(Question: should they be generated in Java? With ANTLR?)
+
+1. Compiler from Python source to AST,
+probably using the PEG parser.
+(If adopting PEG, compile it with CPython and run it with Jython.)
+
+1. Compiler from AST to CPython byte code:
+using the version in Python if possible (compiled with CPython).
+Otherwise, follow CPython implementation in Java.
+(There is no CPython byte code compiler in Jython 2 legacy.)
+
+### Jython Command
+
+1. Jython 3 console command as a Java application
+built over the library.
+
+1. Means to invoke Jython on all major OSes.
+
+### Python `stdlib`
+
+Progressively introduce the `stdlib` and its regression test suite.
+
+The CPython regression test suite is hugely useful
+for driving our conformance and completeness,
+but the test process itself relies on a large proportion
+of the language and `stdlib` working already.
+
+## Implementation Notes
+
+A few principles, some drawn from discussions on `jython-dev` or off-list.
+
+* Adopt or write a higher proportion of modules in Python vs Java,
+  than is the case in Jython 2.
+  (Decent but not high performance interpreter is a pre-requisite.)
+  * Only performance-critical modules are hand-crafted in Java.
+  * Modules that take advantage of Java libraries call them from Python.
+  * Enthusiasm for writing in Java should be directed to re-implementing
+    popular libraries that broaden Jython use (`numpy`, etc.).
+
+* Use fewer external JARs than in Jython 2.
+  * Purpose of each JAR should be documented in the build.
+  * Avoid libraries that circumvent the JVM:
+    * Incompatibility in dealing strings and i/o has been expensive.
+    * Many bugs related to `jnr.posix`. (Replace with `java.nio.file`?)
+    * Carefully consider what `os.*` methods we offer and their semantics.
+    * Reconsider `jnr.jffi`. (Remove related `ctypes` support.)
+
+* Use the dynamic language features (at last) starting with `MethodHandle`.
+  * A core implementation closer to modern CPython.
+  * MVP: `MethodHandle` used to fill type slots (concept proven).
+  * When compiling to JVM byte code, create call sites using the same
+    or a related `MethodHandle`.
+
+* Use a PEG parser following GvR's lead.
+  * Gives us a lot for free (but not a whole compiler).
+  * MVP: Still need a Lexer. (ANTLR?)
+  * MVP: Generate the AST classes from ASDL. (Python or ANTLR?)
+  * MVP: Compiler from AST to Python byte code. (Available in Python?)
+  * Compiler from AST to Java byte code. (Ours to write/update.)
+
+* Use the six module during development with a view to ensuring our
+  users can migrate using it. (Suggestion needs exploration.)
+
+
+## Build environment
+
+All this ought to be followed from the start
+in order to maintain acceptable quality (which is MVP).
+
+* The project is homed at GitHub, where 3.8 is a branch.
```

Comment:
Not sure if we should  be trying to cleanup the existing history somehow. The main thing that I would like changed if the history of checked in external libs, This would not be needed with a Gradle build, but keeping it in the history bloats the repo. But its always a trade off when considering rewriting history... https://git-scm.com/docs/git-filter-branch can do this and removing the `extlibs` dir should be enough.

Reply:
It hardly recommends itself on the man-page 😁 I've already lost several weekends to high-stakes git history wrangling.

You can address this with options to clone. This got me just 16MB of download:

    git clone --depth 4 --branch 3.8-scorched-earth https://github.com/jeff5/jython.git jython-scorched

I don't know if there's a way to make something like that the default experience.

Reply:
Afterthought: I guess the concern is for the CI? I think the same thing is possible, and for Travis a depth of 50 is used by default, but worth checking.